### ISSUE TITLE
Specify installation order of charts

### ISSUE THREAD

**Issue by Docteur-RS** [fluxcd/helm-operator#51]
**Describe the feature**

A cool feature would be a file allowing to set the order of installation of charts.  

**Why?**
Because a lot of Helm charts  are cut in half. First half are CRDs and the second half is the chart itself.  
Take istio for example, it has `istio-init` for the CRDs and `istio` for the chart itself. But if you install `istio` before `istio-init` it will crash because it has not found any CRDs.  
Same when installing Prometheus-operator. It fails because of CRDs not being installed correctly by helm. An error which is hard to prevent because its not possible to set the order of installation.

**What would the new user story look like?**
* 1) Create a new yaml file with a list of each charts and the order of installation
**or**
* 1) Add integer in each HelmRelease files to set order of installation
* 2) Use fluxctl as always
* 3) Installation of each charts follows the specified order of installation
* 4) All helm charts are deployed correctly and we all live happily ever after 

Thx !

**Comment by stefanprodan** [comment]
Installing Istio with Flux and Helm Operator is described here https://github.com/stefanprodan/gitops-istio

TLDR:
- extract the Istio CRDs from the init chart with this [script](https://github.com/stefanprodan/gitops-istio/blob/master/scripts/istio-init.sh) 
- add the CRDs and istio-system namespace to git
- add the Istio HelmRelease to git [example](https://github.com/stefanprodan/gitops-istio/blob/master/istio-system/istio.yaml)
- Flux will apply the CRDs before the Istio HelmRelease

**Comment by Docteur-RS** [comment]
**Hum... Thanks for this ! I'll use it !** 👍 

Though retrieving some kind of integer from each HelmRelease to set the order of installation would solve it altogether and not just for the Istio chart. As this kind of architecture tends to happend more and more often in the helm world ! ;-)

**Comment by stefanprodan** [comment]
Enforcing ordering would make the operator unusable for those that have hundreds of release (we have quite of few users doing this). Helm operator is based on the Kubernetes CRD controller pattern where multiple workers (goroutines) run the reconciliation loop in parallel. Without parallelism, reconciling hundreds of release would take tens of minutes if not even hours.

**Comment by Docteur-RS** [comment]
Thx for your answer.  
I totaly get your point of view though I didn't mean for the hierarchical installation to kill parallelism!  

First, the setting doesn't have to be applied to all charts and secondly only a few of them will be delayed by it.  

For instance lets say we have the following charts and see how they would be deployed:

```
Chart A1 => Time T => Chart A2 => Time T+1 => Chart A3  

Chart B1 => Time T => Chart B2  

Chart C1  

Chart D1  

    ...
```

The charts A1, A2 and A3 need to be deployed in this specific order. Same for B1 and B2.  
For C1 and D1 they have nothing to do together and can be deployed concurently.  

**My point is that A1, B1, C1 and D1 can ALL be installed at the same time because the order of installation is horizontal and not vertical.** This way most of the charts would be deployed immediately.

Next, at `Time T` we only have two concurent installations: A2, B2
And finally at `Time T+1`, only A3.

This way we can respect a specific order of installation without the need to install them one after the other.

---

That's all ;-) !  
I'm sure that you've got your hands full so I won't bother you anymore with this. I just thought it would be something that I could use these days !
(even though I don't have any ideas of how all of this works internally XD).

**Comment by kav** [comment]
Maybe this is a silly suggestion but why not just attempt to install everything at T, reattempt to install failed at T+1, rinse, and repeat. Even the craziest dependency chain will come out in the wash; so to speak

**Comment by Docteur-RS** [comment]
@kav Thanks for your suggestion :-)  
The problem with rebooting a chart till it works is that sometimes there's no apparent signs of crash on the Kubernetes side.    

I have a few charts that when deployed, crash for network reasons. They try to connect to some other charts that have not been deployed yet.  

Unfortunatly they don't always show up as "0/1" in kubernetes meaning they won't be redeployed.   They only relay on ressources being already present and ready to use.  If not, BAM !

Nevertheless your suggestion would work for most use cases !

**Comment by kav** [comment]
Not to backseat code but have you looked into adding liveness and readiness probes all around? I've found Kubernetes to be quite good at managing the chaos with a tiny bit of info from services on their happiness

**Comment by Docteur-RS** [comment]
@kav Haha yeah ! I know you're right ! Unfortunalty, I don't always have access to what is deployed in the Git repository :-(    
And I don't think that I can add readiness probes from the HelmRelease file directly ? **Then I would have to update the chart files themselves which is kind of anoying when everything is overrided from the HelmRelease file and the chart files are left untouched.**  

Letting people deal with the order of instanciation themselves would resolve that.  

Also I would add that I still got issues with some pretty common charts like Prometheus for example. It seems there's some kind of race condition with the prometheus installed by Istio... With the democratisation of GitOps I thinks we'll see thoses errors more and more often.

**Comment by Docteur-RS** [comment]
Moreover specifying the order of installation would not solve the problem of deleting a chart manualy.   
Flux will recreate it and pre existing ressources could blocked the reinstallation process.   
Rendering the way to specify the installation useless.  
😭

**Comment by lucasvmiguel** [comment]
I totally get @Docteur-RS 's problem, in a perfect cloud platform everything should work just fine (it'll crash, but after a while, everything is gonna be recovered), however, we have to consider millions of applications created before the cloud apps architecture. 

That being said, I'd like to see an ordering chart feature like that.

(I will need to implement this feature anyway, I just think that it makes sense to be implemented in the helm operator)

**Comment by renypaul** [comment]
@lucasvmiguel Do you have any plans to work on this in near future ?

**Comment by lucasvmiguel** [comment]
@renypaul 
not really, life has been pretty busy =(

**Comment by renypaul** [comment]
Thanks for the update @lucasvmiguel

**Comment by stefanprodan** [comment]
Install order has been implemented in helm-controller see https://toolkit.fluxcd.io/components/helm/helmreleases/#helmrelease-dependencies
